external/loaders/loaders/mappers/_fine_res.py

```diff
@@ -0,0 +1,241 @@
+    apparent_sources_only = 1
+    apparent_sources_plus_nudging_tendencies = 2
+    apparent_sources_plus_dynamics_differences = 3
+    apparent_sources_extend_lower = 4
+
+
+def _compute_budget(
+    merged: xr.Dataset, approach: Approach, include_temperature_nudging: bool
+) -> MLTendencies:
+
+    merged["Q1"], merged["Q2"] = compute_fine_res_sources(
+        merged, include_temperature_nudging
+    )
+
+    if approach == Approach.apparent_sources_plus_nudging_tendencies:
+        merged["Q1"], merged["Q2"] = _add_nudging_tendencies(merged)
+    elif approach == Approach.apparent_sources_plus_dynamics_differences:
+        merged["Q1"], merged["Q2"] = _add_dynamics_differences(merged)
+    elif approach == Approach.apparent_sources_extend_lower:
+        merged["Q1"] = _extend_lower(merged["Q1"])
+        merged["Q2"] = _extend_lower(merged["Q2"])
+
+    return _ml_standard_names(merged)
+
+
+def _add_nudging_tendencies(merged: xr.Dataset):
+    with xr.set_options(keep_attrs=True):
+        Q1 = merged.Q1 + merged.air_temperature_tendency_due_to_nudging
+        Q2 = merged.Q2 + merged.specific_humidity_tendency_due_to_nudging
+    Q1.attrs.update(
+        {
+            "long_name": merged.Q1.attrs.get("long_name")
+            + " plus dynamics nudging tendencies",
+            "description": merged.Q1.attrs.get("description")
+            + " + dynamics nudging tendencies",
+        }
+    )
+    Q2.attrs.update(
+        {
+            "long_name": merged.Q2.attrs.get("long_name")
+            + " plus dynamics nudging tendencies",
+            "description": merged.Q2.attrs.get("description")
+            + " + dynamics nudging tendencies",
+        }
+    )
+    return Q1, Q2
+
+
+def _add_dynamics_differences(merged: xr.Dataset):
+    with xr.set_options(keep_attrs=True):
+        Q1 = (
+            merged.Q1
+            + merged.fine_minus_coarse_tendency_of_air_temperature_due_to_dynamics
+        )
+        Q2 = (
+            merged.Q2
+            + merged.fine_minus_coarse_tendency_of_specific_humidity_due_to_dynamics
+        )
+    Q1.attrs.update(
+        {
+            "long_name": merged.Q1.attrs.get("long_name")
+            + " plus dynamics differences",
+            "description": merged.Q1.attrs.get("description")
+            + " + dynamics differences",
+        }
+    )
+    Q2.attrs.update(
+        {
+            "long_name": merged.Q2.attrs.get("long_name")
+            + " plus dynamics differences",
+            "description": merged.Q2.attrs.get("description")
+            + " + dynamics differences",
+        }
+    )
+    return Q1, Q2
+
+
+def _extend_lower(fine_source: xr.DataArray, vertical_dim: str = "z") -> xr.Dataset:
+    if fine_source.sizes[vertical_dim] < 2:
+        raise ValueError("vertical_dim must be greater than 1.")
+    fine_source_new_bottom = fine_source.isel({vertical_dim: -2})
+    fine_source_without_bottom = fine_source.isel({vertical_dim: slice(None, -1)})
```

Comment:
my gut would be to use the 3rd level to overwrite the bottom two. Even the 2nd level will be contaminated by the surface level coarsening effects because we do a vertical interpolation during eddy flux convergence calculation.

Reply:
Sure, happy to make that change.